man/man1/raxpol_dat.1
.\" 
.\" Copyright (c) 2016, Gordon D. Carrie. All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 
.\"     * Redistributions of source code must retain the above copyright
.\"     notice, this list of conditions and the following disclaimer.
.\"     * Redistributions in binary form must reproduce the above copyright
.\"     notice, this list of conditions and the following disclaimer in the
.\"     documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
.\" TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
.\" PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
.\" NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
.\" SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\" 
.\" Please address questions and feedback to dev0@trekix.net
.\"
.Pp
.Dd $Mdocdate$
.Dt RAXPOL_DAT 1
.Os UNIX
.Sh NAME
.Nm raxpol_dat
.Nd Print ray data from a RaXPol moment file to standard output.
.Sh SYNOPSIS
.Nm raxpol_dat
.Op Fl V
.Op Fl l
.Op Fl b
.Op Fl l
.Op Fl r
.Op Fl h Ar angle
.Op Fl m Ar moment_name,moment_name,...
.Op Fl s Ar index
.Op Fl c Ar count
.Op Ar raxpol_file
.Sh DESCRIPTION
This application prints ray data from a RaXPol moment file. If
.Ar raxpol_file
is absent or
.Ql - ,
read standard input.
.Pp
The following options are recognized:
.Bl -tag -width DS
.It Fl V
Print version information and exit.
.It Fl r
Print ray headers with the data. Default is to just print ray data.
.It Fl b
Send native memory encoded output instead of ASCII.
.It Fl l
Input file is "old" (2011) format.
.It Fl h
Sets heading, overriding GPS heading in ray headers.
.It Fl m Ar moment_name,moment_name,...
Print data only for moments named in comma separated list.
Each
.Ar moment_name
must be one of
.Bd -literal -offset indent
\fBDBMHC\fP \fBDBMVC\fP \fBDBZ\fP \fBDBZ1\fP \fBVEL\fP \fBZDR\fP \fBPHIDP\fP \fBRHOHV\fP \fBSTD\fP \fBSNRHC\fP \fBSNRVC\fP.

.Ed
Default is to print all moments.
.It Fl s Ar index
Index of first ray to write. First ray is 0.
.It Fl c Ar count
Number of rays to print. Default is to print all rays.
.El
.Sh OUTPUT FORMAT
Default output is ASCII. For each ray, output is:
.Bd -literal -offset indent
  ray_header			(optional)
  moment_name value value ...
  moment_name value value ...
  ...
.Ed
Ray header format is described in
.Xr raxpol_ray_hdrs 1 .
With
.Fl b
option, all output uses native memory encoding. Ray headers, if requested, are
given as anonymous values in the same order as ASCII output. Then for each ray
and moment, process prints the number of bytes in the moment name, followed by
the moment name, followed by data values for all gates. File header provides
the number of gates.
.Sh ENVIRONMENT
Non-zero
.Ev RAXPOL_OLD_FMT
environment variable is same as
.Fl l .
.Sh AUTHOR
.An Gordon Carrie
.Mt dev0@trekix.net


RELEASE_NOTES
Wed Feb 11 16:19:57 CST 2015
raxpol_ray_hdrs also prints file offset to ray

20150316161719
Re-wrote findswps.
Ray header printout includes date as well as time

20150428170551
findswps --
min_span argument can take separate values for PPI and RHI.
Cleaned out dead and deprecated code from findswps.c.
Reformatted output of findswps command for easier searching.
--
raxpol_lib --
Added blocks for 2011 and older raxpol files.
Reformatted output of header print functions for better compatibility with awk.
Added ability to read SPP Sum Power rays.
Various bug fixes. More checking for NULL pointers in functions.
--
raxpol_file_hdr --
prints headers for first and last ray.
Accepts -o option indicating old format files.
--
raxpol_ray_hdr --
Added options.  -a requests abbreviated output.  -h sets heading, overriding
GPS heading in ray headers.  -o input file is "old" (2011) format. Start and
end ray times desired are now specified with -s and -e options, respectively.
--
raxpol_to_sweep --
Accepts separate minimum spans for PPI and RHI.
Is more organinized.
--
tm_calc library has a resolution parameter
raxpol_to_dorade.awk checked in. Makes partial DORADE sweep right now.
raxpol_ray_hdrs utility accepts a time range.

20150507150001
Added RaXPol_Noise function.
Split RaXPol_ZH and RaXPol_ZV into RaXPol_ZH, RaXPol_ZV, RaXPol_DBZH,
and RaXPol_DBZV.
raxpol_file_hdr utility prints offsets to first and last rays.
raxpol_lib --
Created separate raxpol_z and raxpol_dbz functions instead of doing everything
in raxpol_z.
--
raxpol_dat --
utility only calculates and prints requested fields.
raxpol_dat computes moments locally instead of using functions in raxpol_lib,
which are now redundant. RaXPol moments have too many interdependencies for
modular calculations in the current set up.
--

20150925120929
strlcpy.h --
    Disabled global strlcpy before local declaration. This prevents interference
    from Apple's version of strlcpy in their stupid "standard" library.
--
raxpol.h, raxpol_lib.h, raxpol_dat.c --
    Encapsulated the various server modes and recorded moments into RaXPol_Data
    struct.  Consolidated the input and output functions into a smaller set that
    interfaces with the struct. Rewrote raxpol_dat.c with the new interface.
--

20151005153302
findswps.c --
    Added more tests to still_rhi and still_ppi functions.
    Capitalized file scope variables.
--
raxpol.h, raxpol_lib.c, raxpol_dat.c --
    Functions that convert input fields to output are now methods of RaXPol_Data
    struct instead of global funtions.
--

20151005170735
findswps.c --
    Some global variables are now local.
--

20151006200255
raxpol_lib.c --
    Filled in stub functions for single pulse pair and single pulse pair sum
    power output fields.
--

20151013183955
Checking in sweep_img.c, raxpol_swp_idx, raxpol_swp_dat, and raxpol_swp_dat.awk.
Removed raxpol_to_sweep, replace with raxpol_swp_idx.
Checking in geog_proj.c geog_proj.h get_colors.c get_colors.h.
raxpol_lib.c --
    Abbreviated ray header output now includes truck heading.
--
raxpol_dat.c --
    Application no longer prints file header. Printing ray headers is optional.
    Moment names are printed only if there is more than one moment.
--
findswps.c --
    First sweep is index 0, not 1;
--

20151014134133
geog_proj.c --
    GeogProjLonLatToXY and GeogProjXYToLonLat set return arguments to NAN on
    failure.
--
sweep_img.c --
    Added functions to compute gate corners for filled sweeps.
--

20151014190433
findswps, raxpol_swp_idx --
    Now accepts -y option to specify scan type sought, "ppi", "rhi", or "all".
    raxpol_swp_idx uses it.
--

20151102113355
Checking in geog, tm_calc, pisa, color_legend, raxpol_swps_svg, sweep_limits
applications.
Checking in str library utility.
raxpol_ray_hdrs --
    Added ray count option
--
findswps --
    -y option is now -i. This allows the option to pass in
    raxpol_swps_svg, where -y option specifies a style sheet.
--
raxpol_swp_dat, raxpol_swp_dat.awk, raxpol_swp_idx --
    Removed. Not needed.
--
sweep_img --
    Keys end with ":", like pisa.
--
raxpol_lib.c --
    Abbreviated ray header output includes Julian ray time in addition to
    calendar ray time.
--

20151102171827
raxpol_ray_hrs --
    Just takes ray start index and count, or just prints all rays.
    Time search will be a separate application.
--
raxpol_lib.c --
    Shortened abbreviated header output.
--

20151103180433
raxpol_dat --
    Replaced offset argument with index to first ray.
--
sweep_img --
    (bug fix) Corrected order in which gate corners are printed.
    (bug fix) Failure to convert degrees to radians.
--
raxpol_swps_svg --
    Command line options can specify separate bounds for PPI and RHI.
    Adjusted for new abbreviated header output.
--

20151104170343
unix_defs.h --
    c99 is optional.
--
Makefile, raxpol_script_links.svg --
    Path to share directory is substituted into svg script links file.
--

20151104182919
Checking in share directory and contents.
raxpol_swps_svg --
    Color file is now a color directory. Application searches this directory
    for a file with name based on the data type.
    Several variables can now be set either with environment variable or
    command line option.
--

20151105134619
raxpol_dat --
    Always prints moment names, even for only one moment.
--
raxpol_swps_svg --
    Clarified documentation. Strip out moment name from raxpol_dat output.
    sweep_img does not need it.
--

20151105195015
raxpol_dat --
    Now requires starting ray index and ray count. Output consists of moment name
    followed by data for ray count rays, then next moment name followed by data
    for ray count rays, and so on.
--
raxpol_swps_svg --
    Reads the revised raxpol_dat output. Makes sweeps for several moments in
    one block. Desired performance boost failed to show.
--

20151105195715
raxpol_dat, raxpol_swps_svg --
    Restoring from 20151105134619 release. It is simpler and performs as well

src/raxpol_dat.c
/*
   -	raxpol_dat.c --
   -		Write raxpol ray data. See raxpol_dat (1).
   -
   .	Copyright (c) 2014, Gordon D. Carrie. All rights reserved.
   .	
   .	Redistribution and use in source and binary forms, with or without
   .	modification, are permitted provided that the following conditions
   .	are met:
   .	
   .	    * Redistributions of source code must retain the above copyright
   .	    notice, this list of conditions and the following disclaimer.
   .
   .	    * Redistributions in binary form must reproduce the above copyright
   .	    notice, this list of conditions and the following disclaimer in the
   .	    documentation and/or other materials provided with the distribution.
   .	
   .	THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
   .	"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
   .	LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
   .	A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
   .	HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
   .	SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
   .	TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
   .	PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
   .	LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
   .	NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
   .	SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
   .
   .	Please send feedback to dev0@trekix.net
 */

#include "unix_defs.h"
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <limits.h>
#include "alloc.h"
#include "tm_calc_lib.h"
#include "raxpol.h"

static char *argv0;			/* Name of the executable */

#define FFMT "%.5g "			/* Float format */

#define NUM_OUT_MAX 11			/* Number of output fields */

#define RAXPOL_OLD_FMT "RAXPOL_OLD_FMT"

/* Local functions */ 
static void fprintf_field(float *, size_t, char *m, FILE *);
static void fwrite_field(float *, size_t, char *, FILE *);
static float *alloc_field_f(char *, size_t);

int main(int argc, char *argv[])
{
    char *argv0;			/* argv[0] */
    extern char *optarg;		/* See getopt (3) */
    extern int optind;			/* See getopt (3) */
    int c;				/* Index into argv */
    int ray_hdrs = 0;			/* If true, print ray headers, too */
    long r0, num_rays;			/* First ray, number of rays to print */
    long num_rays_max;			/* Number of rays from r0 to EOF */
    char *raxpol_fl_nm;			/* RaXPol file path */
    FILE *raxpol_fl;			/* RaXPol file */
    double dflt_hdg;			/* Default heading */
    long r;				/* Ray index in loop */
    struct RaXPol_Ray_Hdr ray_hdr;	/* Ray header */ 
    off_t o0;				/* Offset to start of first ray */
    off_t o;				/* Offset somewhere in raxpol_fl */
    off_t ray_sz;			/* Size in file of one ray */

    int num_gates;			/* Number of gates */
    struct RaXPol_Data dat;		/* Data for one ray */

    /*
       out_dat = array of all moment names, moment arrays, and functions to
       compute the moments from input ray data.
     */

    struct {
	char *nm;
	float *f;
	int (*calc)(struct RaXPol_Data *, float *);
    } out_dat[NUM_OUT_MAX] = {
	{"DBMHC", NULL, NULL},
	{"DBMVC", NULL, NULL},
	{"DBZ", NULL, NULL},
	{"DBZ1", NULL, NULL},
	{"VEL", NULL, NULL},
	{"ZDR", NULL, NULL},
	{"PHIDP", NULL, NULL},
	{"RHOHV", NULL, NULL},
	{"STD", NULL, NULL},
	{"SNRHC", NULL, NULL},
	{"SNRVC", NULL, NULL}
    };

    int n;				/* Index from out_dat */
    size_t num_out = NUM_OUT_MAX;	/* Number of output moments */
    char *oa;				/* Point into opt_arg */
    char *optarg_m;			/* List of moments from command line */
    char *mm; 				/* Point into optarg_m */
    size_t len;				/* strlen(optarg_m) */
    char *all_moments
	= "DBMHC DBMVC DBZ DBZ1 VEL ZDR PHIDP RHOHV STD SNRHC SNRVC";

    /* Field output */
    int (*prhdr)(struct RaXPol_Ray_Hdr *, FILE *);
    void (*prfld)(float *, size_t, char *m, FILE *);

    argv0 = argv[0];
    r0 = 0;
    num_rays = LONG_MAX;
    prhdr = RaXPol_FPrint_Ray_Hdr;
    prfld =  fprintf_field;
    if ( atoi(getenv(RAXPOL_OLD_FMT) ? getenv(RAXPOL_OLD_FMT) : "0") ) {
	RaXPol_Old_Fmt();
    }
    while ((c = getopt(argc, argv, ":Vblrh:m:s:c:")) != -1) {
	switch(c) {
	    case 'V':
		printf("%s\n", RAXPOL_MOMENT_VERSION);
		exit(EXIT_SUCCESS);
		break;
	    case 'b':
		prhdr = RaXPol_FWrite_Ray_Hdr;
		prfld =  fwrite_field;
		break;
	    case 'r':
		ray_hdrs = 1;
		break;
	    case 'l':
		RaXPol_Old_Fmt();
		break;
	    case 'h':
		if ( sscanf(optarg, "%lf", &dflt_hdg) != 1 ) {
		    fprintf(stderr, "%s: expected float value for default "
			    "heading, got %s\n", argv0, optarg);
		    exit(EXIT_FAILURE);
		}
		RaXPol_Set_Hdg(dflt_hdg);
		break;
	    case 'm':
		/*
		   User has requested a list of moments from command line.
		   Copy optarg to optarg_m, replacing ',' separators with
		   nul string terminators. Assign the resulting strings
		   in optarg_m to the nm members in out_dat.
		 */

		len = strlen(optarg);
		if ( !(optarg_m = CALLOC(len + 1, 1)) ) {
		    fprintf(stderr, "%s: could not allocate storage for moment "
			    "names.\n", argv0);
		    exit(EXIT_FAILURE);
		}
		for (oa = optarg, mm = optarg_m; *oa; oa++, mm++) {
		    *mm = ( *oa == ',' ) ? '\0' : *oa;
		}
		for (n = 0, mm = optarg_m;
			n < NUM_OUT_MAX && mm < optarg_m + len;
			mm++) {
		    if ( *mm ) {
			/* Found start of a moment name in optarg_m */
			if ( !strstr(all_moments, mm) ) {
			    fprintf(stderr, "%s: unknown moment %s. "
				    "Moment must be one of %s\n",
				    argv0, mm, all_moments);
			    exit(EXIT_FAILURE);
			}
			out_dat[n++].nm = mm;

			/* Go to nul terminator at the end of moment name */
			for (++mm; *mm; ++mm) {
			}
		    }
		}
		if ( (num_out = n) == NUM_OUT_MAX ) {
		    fprintf(stdout, "%s: too many moments (%zu). Maximum "
			    "number is %d\n", argv0, num_out, NUM_OUT_MAX);
		    exit(EXIT_FAILURE);
		}
		break;
	    case 's':
		if ( sscanf(optarg, "%ld", &r0) != 1 ) {
		    fprintf(stderr, "%s: expected integer for index of first "
			    "ray, got %s\n", argv0, optarg);
		    exit(EXIT_FAILURE);
		}
	    case 'c':
		if ( sscanf(optarg, "%ld", &num_rays) != 1 ) {
		    fprintf(stderr, "%s: expected integer for count, got %s\n",
			    argv0, optarg);
		    exit(EXIT_FAILURE);
		}
		break;
	    case '?':
		fprintf(stderr, "Usage: %s [-b] [-l] [-h angle] [-s start] "
			"[-c count] [raxpol_file]\n", argv0);
		exit(EXIT_FAILURE);
		break;
	}
    }
    if ( optind ==  argc ) {
	raxpol_fl_nm = "-";
    } else if ( optind + 1 == argc ) {
	raxpol_fl_nm = argv[optind];
    } else {
	fprintf(stderr, "Usage: %s [-b] [-l] [-r] [-h angle] "
		"[-m moment,moment,...] [-s start] [-c count] [raxpol_file]",
		argv0);
	exit(EXIT_FAILURE);
    }
    if ( strcmp(raxpol_fl_nm, "-") == 0 ) {
	raxpol_fl = stdin;
    } else if ( !(raxpol_fl = fopen(raxpol_fl_nm, "r")) ) {
	fprintf(stderr, "%s: could not open %s for reading.\n",
		argv0, raxpol_fl_nm);
	exit(EXIT_FAILURE);
    }

    /* Read file header. Compute some convenience constants. */ 
    if ( !RaXPol_Init_Data(&dat, raxpol_fl) ) {
	fprintf(stderr, "%s: failed to initialize RaXPol data structure.\n",
		argv0);
	exit(EXIT_FAILURE);
    }
    num_gates = dat.file_hdr.num_rng_gates;
    if ( (o0 = ftello(raxpol_fl)) == -1 ) {
	fprintf(stderr, "%s: could not determine position in file.\n%s\n",
		argv0, strerror(errno));
	exit(EXIT_FAILURE);
    }

    /* Allocate requested output fields and assign associated calc members. */
    for (n = 0; n < num_out; n++) {
	if ( strcmp(out_dat[n].nm, "DBMHC") == 0 ) {
	    out_dat[n].f = alloc_field_f("DBMHC", num_gates);
	    out_dat[n].calc = dat.dbmhc;
	} else if ( strcmp(out_dat[n].nm, "DBMVC") == 0 ) {
	    out_dat[n].f = alloc_field_f("DBMVC", num_gates);
	    out_dat[n].calc = dat.dbmvc;
	} else if ( strcmp(out_dat[n].nm, "DBZ") == 0 ) {
	    out_dat[n].f = alloc_field_f("DBZ", num_gates);
	    out_dat[n].calc = dat.dbz;
	} else if ( strcmp(out_dat[n].nm, "DBZ1") == 0 ) {
	    out_dat[n].f = alloc_field_f("DBZ1", num_gates);
	    out_dat[n].calc = dat.dbz1;
	} else if ( strcmp(out_dat[n].nm, "VEL") == 0 ) {
	    out_dat[n].f = alloc_field_f("VEL", num_gates);
	    out_dat[n].calc = dat.vel;
	} else if ( strcmp(out_dat[n].nm, "ZDR") == 0 ) {
	    out_dat[n].f = alloc_field_f("ZDR", num_gates);
	    out_dat[n].calc = dat.zdr;
	} else if ( strcmp(out_dat[n].nm, "PHIDP") == 0 ) {
	    out_dat[n].f = alloc_field_f("PHIDP", num_gates);
	    out_dat[n].calc = dat.phidp;
	} else if ( strcmp(out_dat[n].nm, "RHOHV") == 0 ) {
	    out_dat[n].f = alloc_field_f("RHOHV", num_gates);
	    out_dat[n].calc = dat.rhohv;
	} else if ( strcmp(out_dat[n].nm, "STD") == 0 ) {
	    out_dat[n].f = alloc_field_f("STD", num_gates);
	    out_dat[n].calc = dat.std;
	} else if ( strcmp(out_dat[n].nm, "SNRHC") == 0 ) {
	    out_dat[n].f = alloc_field_f("SNRHC", num_gates);
	    out_dat[n].calc = dat.snrhc;
	} else if ( strcmp(out_dat[n].nm, "SNRVC") == 0 ) {
	    out_dat[n].f = alloc_field_f("SNRVC", num_gates);
	    out_dat[n].calc = dat.snrvc;
	}
    }

    /* Determine ray size and num_rays_max */
    if ( !RaXPol_Read_Ray_Hdr(&ray_hdr, raxpol_fl) ) {
	fprintf(stderr, "%s: failed to read header for first ray.\n",
		argv0);
	exit(EXIT_FAILURE);
    }
    if ( (o = ftello(raxpol_fl)) == -1 ) {
	fprintf(stderr, "%s: could not determine position in file.\n%s\n",
		argv0, strerror(errno));
	exit(EXIT_FAILURE);
    }
    ray_sz = o - o0 + ray_hdr.data_size;
    if ( fseeko(raxpol_fl, 0, SEEK_END) == -1
	    || (o = ftello(raxpol_fl)) == -1 ) {
	fprintf(stderr, "%s: could not position at end of file.\n"
		"%s\n", argv0, strerror(errno));
	exit(EXIT_FAILURE);
    }
    num_rays_max = (o - o0) / ray_sz - r0;
    if ( num_rays > num_rays_max ) {
	num_rays = num_rays_max;
    }

    /* Move to first ray */
    if ( r0 > 0 ) {
	if ( fseeko(raxpol_fl, o0 + r0 * ray_sz, SEEK_SET) == -1 ) {
	    fprintf(stderr, "%s: could not position at start of first ray.\n"
		    "%s\n", argv0, strerror(errno));
	    exit(EXIT_FAILURE);
	}
    } else {
	if ( fseeko(raxpol_fl, o0, SEEK_SET) == -1) {
	    fprintf(stderr, "%s: could not position at end of file.\n"
		    "%s\n", argv0, strerror(errno));
	    exit(EXIT_FAILURE);
	}
    }

    /* Read and print rays. */
    for (r = r0; r < r0 + num_rays; r++) {
	if ( !RaXPol_Read_Ray(&dat, raxpol_fl) ) {
	    if ( ferror(raxpol_fl) ) {
		fprintf(stderr, "%s: could not read ray %ld\n",
			argv0, r);
		exit(EXIT_FAILURE);
	    } else {
		exit(EXIT_SUCCESS);
	    }
	}
	if ( ray_hdrs ) {
	    if ( prhdr == RaXPol_FPrint_Ray_Hdr ) {
		printf("ray %ld\n", r);
	    }
	    prhdr(&ray_hdr, stdout);
	}
	for (n = 0; n < num_out; n++) {
	    if ( !out_dat[n].calc(&dat, out_dat[n].f) ) {
		fprintf(stderr, "%s: could not compute %s for ray %ld\n",
			argv0, out_dat[n].nm, r);
		exit(EXIT_FAILURE);
	    }
	    prfld(out_dat[n].f, num_gates, out_dat[n].nm, stdout);
	}
	printf("\n");
    }

    return 0;
}

/*
   If not "", print nm to out. Then print n ascii floats from f to out,
   followed by a newline.
 */

static void fprintf_field(float *f, size_t n, char *nm, FILE *out)
{
    float *f1;				/* End of array f */

    if ( !f || !nm ) {
	fprintf(stderr, "No data to print for field %s\n", nm ? nm : "unknown");
	exit(EXIT_FAILURE);
    }
    if ( strlen(nm) > 0 ) {
	fprintf(out, "%s ", nm);
    }
    for (f1 = f + n; f < f1; f++) {
	fprintf(out, FFMT, *f);
    }
    fprintf(out, "\n");
}

/* Write n floats from f to out. Field assumed to be named nm. */ 
static void fwrite_field(float *f, size_t n, char *nm, FILE *out)
{
    size_t sz = strlen(nm);

    if ( !f || !nm ) {
	fprintf(stderr, "No data to print for field %s\n", nm ? nm : "unknown");
	exit(EXIT_FAILURE);
    }
    if ( fwrite(&sz, sizeof(size_t), 1, out) != 1 ) {
	fprintf(stderr, "%s: could not write size of field name for %s.\n",
		argv0, nm);
	exit(EXIT_FAILURE);
    }
    if ( fwrite(nm, sz, 1, out) != 1 ) {
	fprintf(stderr, "%s: could not write field name for %s.\n", argv0, nm);
	exit(EXIT_FAILURE);
    }
    if ( fwrite(f, sizeof(float), n, out) != n ) {
	fprintf(stderr, "%s: could not write field name for %s.\n", argv0, nm);
	exit(EXIT_FAILURE);
    }
}

/* Allocate memory for an output field named nm with space for n floats */ 
static float *alloc_field_f(char *nm, size_t n)
{
    float *dat;

    if ( !(dat = calloc(n, sizeof(float))) ) {
	fprintf(stderr, "%s: could not allocate %u values for %s\n",
		argv0, (unsigned)n, nm);
	exit(EXIT_FAILURE);
    }
    return dat;
}

